src/payment/strategies/googlepay/googlepay-payment-strategy.ts

```diff
@@ -0,0 +1,181 @@
+import { PaymentStrategy } from '../';
+import {
+    Payment,
+    PaymentActionCreator,
+    PaymentInitializeOptions,
+    PaymentMethodActionCreator,
+    PaymentRequestOptions,
+    PaymentStrategyActionCreator
+} from '../..';
+import { CheckoutActionCreator, CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
+import { NotInitializedError } from '../../../common/error/errors';
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType,
+    NotInitializedErrorType,
+} from '../../../common/error/errors';
+import { toFormUrlEncoded } from '../../../common/http-request';
+import { bindDecorator as bind } from '../../../common/utility';
+import {
+    OrderActionCreator, OrderRequestBody } from '../../../order';
+
+import { GooglePayPaymentInitializeOptions, GooglePayPaymentProcessor } from './';
+import {
+    GooglePaymentData,
+    GooglePayAddress,
+    GooglePayInitializer,
+    PaymentMethodData,
+    TokenizePayload,
+} from './googlepay';
+
+export default class GooglePayPaymentStrategy extends PaymentStrategy {
+    private _googlePayOptions!: GooglePayPaymentInitializeOptions;
+    private _methodId!: string;
+    private _walletButton?: HTMLElement;
+
+    constructor(
+        store: CheckoutStore,
+        private _checkoutActionCreator: CheckoutActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _paymentStrategyActionCreator: PaymentStrategyActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _orderActionCreator: OrderActionCreator,
+        private _googlePayInitializer: GooglePayInitializer,
+        private _requestSender: RequestSender,
+        private _googlePayPaymentProcessor: GooglePayPaymentProcessor
+    ) {
+        super(store);
+    }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        this._methodId = options.methodId;
+
+        if (!options.googlepay) {
+            throw new InvalidArgumentError('Unable to initialize payment because "options.googlepay" argument is not provided.');
+        }
+
+        this._googlePayOptions = options.googlepay;
+
+        const walletButton = options.googlepay.walletButton && document.getElementById(options.googlepay.walletButton);
+
+        if (walletButton) {
+            this._walletButton = walletButton;
+            this._walletButton.addEventListener('click', this._handleWalletButtonClick);
+        }
+
+        return this._googlePayPaymentProcessor.initialize(this._methodId)
+            .then(() => super.initialize(options));
+    }
+
+    deinitialize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        if (this._walletButton) {
+            this._walletButton.removeEventListener('click', this._handleWalletButtonClick);
+        }
+
+        this._walletButton = undefined;
+
+        return Promise.all([
+            this._googlePayInitializer.teardown(),
```

Comment:
Is there a reason why you need to call `teardown` directly? I thought `GooglePayPaymentProcessor#deinitialize ` should handle it already.